api/cryptocom/ws_actions_test.go

```diff
@@ -0,0 +1,568 @@
+package cryptocom
+
+import (
+	"bytes"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/bmizerany/assert"
+	"github.com/google/uuid"
+	"github.com/shopspring/decimal"
+)
+
+type mockRequest struct {
+	ID     int    `json:"id"`
+	Method string `json:"method"`
+	Nonce  string `json:"nonce"`
+	Params map[string]interface{}
+}
+
+type testingFunc func(client *Client)
+
+func TestFormat(t *testing.T) {
+	client := New("", "", "test", "test")
+
+	markets := []string{"ETH_BTC", "ETH_COV", "XRP_BTC"}
+	expected := []string{"trade.ETH_BTC", "trade.ETH_COV", "trade.XRP_BTC"}
+
+	result := client.format(markets, func(s string) string {
+		return fmt.Sprintf("trade.%s", s)
+	})
+
+	assert.Equal(t, result, expected)
+}
+
+func testSubscribe(t *testing.T, expected string, isPrivate bool, testFunc testingFunc) {
+	// prepare expected
+	var expectedResponse mockRequest
+	err := json.Unmarshal([]byte(expected), &expectedResponse)
+	if err != nil {
+		t.Fatal("error on parse expected")
+	}
+
+	// prepare mock
+	client := New("test", "test", "test", "test")
+	privateWritingMessage := bytes.NewBuffer(nil)
+	publicWritingMessage := bytes.NewBuffer(nil)
+	client.connectMock(bytes.NewBuffer(nil), bytes.NewBuffer(nil), privateWritingMessage, publicWritingMessage)
+
+	// call test function
+	testFunc(client)
+
+	// get response
+	var writingMessage mockRequest
+	if isPrivate {
+		err = json.Unmarshal(privateWritingMessage.Bytes(), &writingMessage)
+	} else {
+		err = json.Unmarshal(publicWritingMessage.Bytes(), &writingMessage)
+	}
+	if err != nil {
+		t.Fatal("error on parse writing message")
+	}
+
+	// assertion
+	assert.NotEqual(t, mockRequest{}, writingMessage)
+	// doesn't assert on nonce
+	assert.Equal(t, expectedResponse.ID, writingMessage.ID)
+	assert.Equal(t, expectedResponse.Method, writingMessage.Method)
+	assert.Equal(t, expectedResponse.Params, writingMessage.Params)
+}
+
+// for test parse json. this case expected and mock is the same thing
+func testResponse(t *testing.T, expected string, isPrivate bool) {
+	// prepare expected
+	var expectedResponse Response
+	err := json.Unmarshal([]byte(expected), &expectedResponse)
+	if err != nil {
+		t.Fatal("error on parse expected")
+	}
+
+	// prepare mock
+	client := New("test", "test", "test", "test")
+	var privateResponse *bytes.Buffer
+	var publicResponse *bytes.Buffer
+
+	testID := 9999
+	emptyResponse := fmt.Sprintf(`{"id":%d,"method":"subscribe"}`, testID) // Id 9999 is flag for test
+	// mock response.
+	if isPrivate {
+		privateResponse = bytes.NewBufferString(expected)
+		publicResponse = bytes.NewBufferString(emptyResponse)
+	} else {
+		privateResponse = bytes.NewBufferString(emptyResponse)
+		publicResponse = bytes.NewBufferString(expected)
+	}
+	client.connectMock(privateResponse, publicResponse, bytes.NewBuffer(nil), bytes.NewBuffer(nil))
+
+	msgs := client.Listen()
+
+	var response Response
+	for m := range msgs {
```

Comment:
you read any message that does not have id 9999, but why do you need to write that?

Reply:
Because msgs it receives from the public and private response with an infinite loop. But when I test I need to receive only one response(from public or private). Therefore I set the response that I don't want to test to id 9999 for skip a response.